### PR TITLE
feat: Add Initiative and Project Associations

### PR DESCRIPTION
**Pull request by dbouwman**
1. Description:

Key addition is implementation of Project to Initiative "associations"

Finalized terminology:

image

Also exposes functions which build `IQuery` objects to show lists of "accepted", and "pending" Projects, for a given initiative. This can be passed into a Gallery. Functions that executes these queries is also exposed.

There are also internal functions for generating these sort of query objects, as this pattern will be very useful as we lean into using Galleries for lists (vs manual fetching, then mapping results into the DOM)

1. Instructions for testing: run tests

1. Closes Issues: #7779

1. [x] Updated meaningful TSDoc to methods including Parameters and Returns, see [Documentation Guide](https://esri.github.io/hub-components/storybook/?path=/story/guides-documentation--page)

1. [x] used semantic commit messages
  
1. [x] PR title follows semantic commit format (**CRITICAL** if the title is not in a semantic format, the release automation will not run!)

1. [ ] updated `peerDependencies` as needed. **CRITICAL** our automated release system can **not** be counted on to update `peerDependencies` so we _must_ do it manually in our PRs when needed. See the [updating peerDependencies](/RELEASE.md#Updating-peerDependencies) section of the release instructions for more details.

### CODE REVIEW

File: packages/common/src/groups/HubGroups.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { IGroup } from "@esri/arcgis-rest-types";
 import { fetchGroupEnrichments } from "./_internal/enrichments";
 import { getProp, setProp } from "../objects";
-import { getGroupThumbnailUrl, IHubSearchResult } from "../search";
+import { getGroupThumbnailUrl } from "../search/utils";
```
Comment (dbouwman) [inline packages/common/src/groups/HubGroups.ts:4]: Changes needed to make node tests pass...

File: packages/common/src/groups/_internal/computeProps.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { UserSession } from "@esri/arcgis-rest-auth";
 import { IHubGroup } from "../../core/types/IHubGroup";
 import { IGroup } from "@esri/arcgis-rest-types";
 import { isDiscussable } from "../../discussions";
-import { getGroupThumbnailUrl } from "../../search";
+import { getGroupThumbnailUrl } from "../../search/utils";
```
Comment (dbouwman) [inline packages/common/src/groups/_internal/computeProps.ts:7]: Changes needed to make node tests pass...

File: packages/common/src/metrics/resolveMetric.ts [hunk 3]
```diff
@@ -168,10 +168,15 @@ async function resolveItemQueryMetric(
...
+  // Memoization is great but we have UI's where we want immediate
+  // updates as we share more items into the groups, which means
+  // we can't memoize the search response
   // create/get memoized version of portalSearchItemsAsItems
-  const memoizedPortalSearch = memoize(portalSearchItemsAsItems);
+  // const memoizedPortalSearch = memoize(portalSearchItemsAsItems);
   // Execute the query
-  const response = await memoizedPortalSearch(combined, opts);
+  // const response = await memoizedPortalSearch(combined, opts);
+
```
Comment (dbouwman) [inline packages/common/src/metrics/resolveMetric.ts:178]: If this becomes a performance issue, we can look at implementing a means to optionally use memoization

File: packages/common/src/permissions/types/Permission.ts [hunk 4]
```diff
@@ -30,7 +29,15 @@ const validPermissions = [
...
-export type Permission = (typeof validPermissions)[number];
+export type Permission =
+  | (typeof SitePermissions)[number]
+  | (typeof ProjectPermissions)[number]
+  | (typeof InitiativePermissions)[number]
+  | (typeof ContentPermissions)[number]
+  | (typeof GroupPermissions)[number]
+  | (typeof PagePermissions)[number]
+  | (typeof PlatformPermissions)[number]
+  | (typeof TempPermissions)[number];
```
Comment (dbouwman) [inline packages/common/src/permissions/types/Permission.ts:40]: This re-enabled intellisense on the permission string

File: packages/common/src/search/hubSearch.ts [hunk 5]
```diff
@@ -8,14 +8,13 @@ import {
...
-  portalSearchUsers,
-  hubSearchChannels,
-} from "./_internal";
+
 import { portalSearchGroupMembers } from "./_internal/portalSearchGroupMembers";
+import { portalSearchItems } from "./_internal/portalSearchItems";
+import { portalSearchGroups } from "./_internal/portalSearchGroups";
+import { portalSearchUsers } from "./_internal/portalSearchUsers";
+import { hubSearchItems } from "./_internal/hubSearchItems";
+import { hubSearchChannels } from "./_internal/hubSearchChannels";
```
Comment (dbouwman) [inline packages/common/src/search/hubSearch.ts:17]: Changes needed to make node tests pass...

File: packages/common/src/users/HubUsers.ts [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
 import { getProp } from "../objects";
-import { getUserThumbnailUrl, IHubSearchResult } from "../search";
+import { getUserThumbnailUrl } from "../search/utils";
 import { parseInclude } from "../search/_internal/parseInclude";
 import { IHubRequestOptions } from "../types";
 import { getUserHomeUrl } from "../urls";
 import { unique } from "../util";
 import { mapBy } from "../utils";
 import { AccessLevel } from "../core";
+import { IHubSearchResult } from "../search/types/IHubSearchResult";
```
Comment (dbouwman) [inline packages/common/src/users/HubUsers.ts:11]: Changes needed to make node tests pass...

File: packages/common/src/core/HubItemEntity.ts [hunk 7]
```diff
@@ -383,4 +389,32 @@ export abstract class HubItemEntity<T extends IHubItemEntity>
...
   }
+
+  /**
+   * Return a list of IAssociationInfo objects representing
+   * the associations this entity has, to the specified type
+   * @param type
+   * @returns
+   */
+  listAssociations(type: AssociationType): IAssociationInfo[] {
+    return listAssociations(this.entity, type);
```
Comment (tomwayson) [inline packages/common/src/core/HubItemEntity.ts:400]: seems a little odd that the add and remove functions don't follow the signature pattern of list where the entity comes first.

**Review by tomwayson — APPROVED** [review]
LGTM. just a couple thoughts

File: packages/common/src/core/HubItemEntity.ts (hunk 7 above)
Comment (dbouwman) [reply packages/common/src/core/HubItemEntity.ts:400]: good point - when I was writing the tests I had thought about this, but likely got distracted and did not implement. Changed.

File: packages/common/src/associations/internal/getTypeFromAssociationType.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { AssociationType } from "../types";
+
+/**
+ * Get the item type for an association type
+ * @param type
+ * @returns
+ */
+export function getTypeFromAssociationType(type: AssociationType) {
```
[the remark this answers is not in the document]
Comment (dbouwman) [reply packages/common/src/associations/internal/getTypeFromAssociationType.ts:8]: I went back and forth on this, but AT has expressed that in an ideal world we can "associate" any entity to any other - so could be more than item type. I'm not sure we'll end up there, but this leaves it open

**Pull request merged**